File: Day1-9/1.py
def captcha(captcha_input):
    result = 0
    for i in range(len(captcha_input)):
        next_char = captcha_input[(i+1)%len(captcha_input)]
        if captcha_input[i] == next_char:
            result += int(next_char)
    return result


def captcha2(captcha_input):
    result = 0
    n = len(captcha_input)
    for i in range(len(captcha_input)):
        next_char = captcha_input[(i + n//2)%len(captcha_input)]
        if captcha_input[i] == next_char:
            result += int(next_char)
    return result


if __name__ == '__main__':
    with open('1.txt') as f:
        puzzle = f.read().strip()
    assert captcha('1122') == 3
    assert captcha('1111') == 4
    assert captcha('1234') == 0
    assert captcha('91212129') == 9
    part1 = captcha(puzzle)
    print("Part 1: {}".format(part1))

    part2 = captcha2(puzzle)
    print("Part 2: {}".format(part2))

File: Day1-9/1.rs
use std::fs::File;
use std::io::prelude::*;

fn main() {
    let puzzle = get_puzzle_input("1.txt");

    assert!(captcha1("1122") == 3);
    println!("Part 1: {}", captcha1(&puzzle));
    println!("Part 2: {}", captcha2(&puzzle));
}

fn get_puzzle_input(filename: &str) -> String {
    let mut f = File::open(filename).expect("file not found");

    let mut contents = String::new();
    f.read_to_string(&mut contents)
        .expect("something went wrong reading the file");

    contents.replace("\n", "")
}

fn captcha1(input: &str) -> u32 {
    let mut counter = 0;
    let bytes = input.as_bytes();
    for (i, c) in bytes.iter().enumerate() {
        let next_idx = (i + 1) % input.len();
        if bytes[next_idx] == *c {
            counter += (*c as char).to_digit(10).unwrap();
        }
    }
    counter
}

fn captcha2(input: &str) -> u32 {
    let mut counter = 0;
    let bytes = input.as_bytes();
    for (i, c) in bytes.iter().enumerate() {
        let next_idx = (i + input.len() / 2) % input.len();
        if bytes[next_idx] == *c {
            counter += (*c as char).to_digit(10).unwrap();
        }
    }
    counter
}
